meta-balena-common/recipes-support/resin-init/resin-init-flasher/resin-init-flasher

```diff
@@ -133,26 +133,78 @@ if [ -z "$internal_dev" ]; then
     fail "Failed to find any block devices in $INTERNAL_DEVICE_KERNEL."
 fi
 inform "$internal_dev will be used for flashing."
-dd if="/opt/$BALENA_IMAGE" of="/dev/$internal_dev" bs=4M 2> /tmp/dd_progress_log & DD_PID=$!
 
-if ! kill -0 $DD_PID; then
-    fail "Failed to flash internal device $INTERNAL_DEVICE_KERNEL."
-fi
+if [ "$LUKS" = "1" ]; then
+    # TODO: Hardcoded password now, randomize with TPM2
+    PASSWORD="${LUKS_PASSWORD:-balenabeta}"
+    LOOP_DEVICE=$(losetup --find --show --partscan "/opt/$BALENA_IMAGE")
+
+    # Copy header and resin-boot partition
+    inform "Flashing boot partition"
+    dd if="/opt/$BALENA_IMAGE" of="/dev/$internal_dev" bs=4M count=11
+
+    # Repartition the new drive - this assumes GPT partition table
+    inform "Repartitioning $internal_dev for disk encryption"
+
+    echo -e "OK\nFix" | parted ---pretend-input-tty "/dev/$internal_dev" print
+
+    parted -s "/dev/$internal_dev" -- rm 5
+    parted -s "/dev/$internal_dev" -- rm 4
+    parted -s "/dev/$internal_dev" -- rm 3
+    parted -s "/dev/$internal_dev" -- rm 2
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-boot 44 86
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-rootA 86 1112
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-rootB 1112 2138
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-state 2138 2160
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-data 2160 2366
+
+    PART_PREFIX=""
+    if [ -e "/dev/${internal_dev}p1" ]; then
+        PART_PREFIX="p"
+    fi
 
-IMAGE_FILE_SIZE=$(wc -c /opt/"$BALENA_IMAGE" | awk '{print $1}')
+    # Relabel former boot partition
+    parted -s "/dev/$internal_dev" -- name 1 balena-efi
+    mlabel ::balena-efi -i "/dev/${internal_dev}${PART_PREFIX}1"
+
+    for PART_ID in 2 3 4 5 6; do
```

Comment:
Maybe we can calculate how many partition we have instead of hardcoding the iteration. Some DTs have tens of partitions.

Reply:
Changed